### ISSUE TITLE
Question: Is it appropriate to merge styles using javascript? Why or why not?

### ISSUE THREAD

**Issue by joshuarossi** [css-modules/css-modules#309]
Start with the following:

```
/* styles/base.module.css */
.test {
color: red;
}
```

and

```
/* styles/theme.module.css */
.test {
color: orange;
}
```

and

```
/* styles/index.js */
import base from './base.module.css'
import custom from './custom.module.css'

const styles = {...base, ...custom}

export default styles
```

Now, when i 

```
import styles from './styles'
export const Button = props => (<button className={styles.test}>{props.label}</button>)

// The button's text would normally be red, but in this case is overridden by the theme to be orange.
```

I see that my base style is overridden by my theme style. This means that to create a new theme, I simply receive a theme.module.css and add it to the folder, and if it changes something, it can do so while not modifying the rest. 

My question is, how does this compare to composes? A third party would also need to understand the modules syntax. Using the JS approach would allow them to write standard vanilla css. 

Just trying to understand the reasoning behind composes.

**Comment by joshuarossi** [comment]
Sorry, I didn't explain it that well...let's say you have two classes

custom.css has 
p {
background-color: red;
}
base.css
p {
background-color: red;
color: blue;
}

you will lose "color: blue", so you can use it to override an entire class, but you must replace EVERY value or you will not get anything for that value, which is probably ok for most uses. 

I was mainly curious when and where you would bother using "extends" since you can use purely native stuff, and that is the reason i think, that you can't partially compose classes using just object spreading.

**Comment by joshuarossi** [comment]
but in general, I prefer the object spreading approach because it doesn't add non-css stuff to the css file (for portability, etc)

**Comment by Download** [comment]
@joshuarossi Yes agreed, when doing it to override values you need to take care to make lots of classes for single rules so to be able to override them easily.

My use case is different actually. I have a .scss stylesheet in one module:

```scss
.some_class {
  background: green;
}
```

And I have a Preact component in another... The Preact component renders it's HTML using the CSS classes from the .scss stylesheet. It defines those classes in a JS object with the same interface as CSS Modules would create:

```js
const defaultClasses = {
  some_class: 'some_class'
}
```

Then in the application, we use CSS Modules to load the .scss and it creates a module definition that looks like this: 

```js
const cssModule = { 
  some_class: 'some_class_3gf47'
}
```

The added hash makes the style unique so it will not clash. Now we pass the module definition to the Preact component. It uses spread operator to merge the CSS definition over the default object it created:

```js
var classes = { ...defaultClasses, ...cssModule }
```

and ends up rendering the HTML using `some_class_3gf47` as the class name and all works great!

The nice thing is that if the application is not using CSS Modules but only SASS, he just does not pass the CSS Module definition and it will still work.

If you want to see more check these projects where I am using it:
* [solids](https://github.com/download/solids) *scss*
* [preact-solids](https://github.com/download/preact-solids) *preact component*
* [solids-www](https://github.com/download/solids-www) *example application*

Far from done though...

**Comment by ron0115** [comment]
@joshuarossi  HI, I Found that using lodash,merge make it works.
`
    import { mergeWith } from 'lodash'
    const comStyles = mergeWith(styles, defaultStyles, (oldValue, newValue) => {
      return oldValue && oldValue.indexOf(newValue) > -1 ? oldValue : `${oldValue} ${newValue}`
    })
`

**Comment by vasilenka** [comment]
@joshuarossi if what you need is overriding and also combining properties, maybe you can use [classnames](https://github.com/JedWatson/classnames)

My use case usually goes like this

```css
// styles.module.css

.base {
	background-color: red;
	color: coral;
}
.override {
	background-color: blue;
}
```

```javascript
// on your components

import cx from classnames;
import styles from './styles.module.css';

let combined = cx(styles.base, styles.override);

// combined will have:
// { 
//	color: coral; 
//	background-color: blue; 
// }

<Component className={combined} />
```

And there're lot of ways you can combined those styles, and it's not limited to just css class selector.

**Comment by Download** [comment]
@ron0115 That's an interesting approach! You are actually making a list of style classes this way.

**Comment by ron0115** [comment]
> @ron0115 That's an interesting approach! You are actually making a list of style classes this way.

But it couldn't resovle the problem of CSS Priority, because the Priority depends on the order of css in the document
